libs/utils/src/rate_limit_closure.rs

```diff
@@ -0,0 +1,66 @@
+//! A helper to rate limit the invocation of a closure.
+
+use std::time::{Duration, Instant};
+
+pub struct RateLimitClosure {
+    last: Option<Instant>,
+    interval: Duration,
+}
+
+impl RateLimitClosure {
+    pub fn new(interval: Duration) -> Self {
+        Self {
+            last: None,
+            interval,
+        }
+    }
+
+    pub fn call<F: FnOnce()>(&mut self, f: F) {
+        let now = Instant::now();
+        match self.last {
+            Some(last) if now - last > self.interval => {
+                f();
+                self.last = Some(now);
+            }
+            None => {
+                f();
+                self.last = Some(now);
+            }
+            Some(_) => {}
+        }
```

Comment:
otherwise if you really want the positive path only:

```suggestion
        let call_it = self.last.as_ref()
            .map(|x| now - last > self.interval)
            .unwrap_or(true);

        if call_it {
            self.last = Some(now);
            f();
        }
```